0027-safely-recommend-actions.adoc

```diff
@@ -0,0 +1,173 @@
+
+The atomic guarantees of the COMIT protocols rely on the beta expiry time being _before_ the alpha expiry time, these expiry times are absolute (as opposed to relative) i.e., seconds since epoch.
+
+In order for a swap to be guaranteed atomic the redeem transaction must be included in the blockchain _before_ the expiry time elapses. Therefore the redeem action is subject to some `TTI` calculated for the given blockchain. Repeating because this is important; the expiry time represents a hard limit, the redeem transaction must be included in a block before the expiry elapses for the swap to be atomic.
+
+Therefore, there exists a time `T` after which attempting the redeem action is considered 'unsafe' i.e., the probability of the transaction being included in a block is less than some predefined tolerance:
+
+----
+T = <expiry time> - TTI
+----
+
+Please note, that the `TTI` need not be the same for the redeem transaction as for the fund transaction since different fees can be used for each transaction as required.
+
+== Functionality
+
+The system is expected to provide the following functionality:
+
+.Requirements
+1. Given two blockchains, create a pair of useful expiries.
+2. Given a pair of expiries and their respective blockchains, determine if the expiries are useful.
+3. Given the current swap state, determine what the user should do.
+
+A pair of expiries is 'useful' if, at this moment, both parties are able to execute all steps to complete the swap.
+
+== Solution
+
+Here we describe solutions to each of the listed requirements above.
+
+1. To create a pair of useful expiries we calculate the expected time required for completion of the swap. That is the sum of the transition periods for each transition. We then add a Time To Live (`TTL`) value, this ensures the calculated expiries remain useful for `TTL`.
+
+2. A pair of expiries is useful if there is time to take each action required to complete the swap i.e., the expiry pair is useful if there is time to transition to the first action AND there is time for each subsequent action.
+
+3. Semantically the next possible action is either the next happy path action or cancel. Cancel is either refund or abort.
+
+.Semantic actions
+* *next*: The next happy path action (deterministic from the state diagram).
+* *refund*: Send a refund transaction to the blockchain (cancel after fund).
+* *wait_for_refund*: Refund is not yet available but should be taken when available.
+* *abort*: Take no further action (cancel before fund).
+
+The recommended next action can be fully determined from the following algorithm:
+
+----
+1. IF refund is available RETURN refund
+2. IF there is time to transition to next AND there is time for the subsequent action RETURN next
+3. IF have funded RETURN wait_for_refund
+4. RETURN abort
+----
```

Comment:
I find it hard to follow this algorithm. Especially the recursive part of step 2. Maybe a structured [pseudocode](https://en.wikipedia.org/wiki/Pseudocode#Example) approach would more suitable?

Reply:
I used
```
IF refund is available (implies fund has occurred) AND expiry time has elapsed
        RETURN refund

IF there is time to transition from the current state to the next state AND there is time to complete the next action
        RETURN next

IF we have funded
        RETURN wait_for_refund

RETURN abort
```